### ISSUE TITLE
Cannot run Lua-Terra interaction example

### ISSUE THREAD

**Issue by renatoathaydes** [terralang/terra#567]
[This example](https://terralang.org/getting-started.html#lua-terra-interaction):

```lua
function add1(a)
    a.real = a.real + 1
end
struct Complex { real : double, imag : double }
terra doit()
    var a : Complex = {1,2}
    add1(&a)
    return a
end
a = doit()
print(a.real,a.imag) -- 2    1
print(type(a)) -- cdata
```

When I run with the distribution you've just sent me I get a bunch of errors:

```bash
renato@renato-desktop:~/terra-Linux-arm64-7dc387c$ cat a.t
function add1(a)
    a.real = a.real + 1
end
struct Complex { real : double, imag : double }
terra doit()
    var a : Complex = {1,2}
    add1(&a)
    return a
end
a = doit()
print(a.real,a.imag) -- 2    1
print(type(a)) -- cdata

renato@renato-desktop:~/terra-Linux-arm64-7dc387c$ bin/terra a.t
/terra/src/terralib.lua:1172: Errors reported during typechecking.
a.t:6: invalid conversion from {int32,int32} to Complex
    var a : Complex = {1,2}
                      ^
a.t:7: lua object of type function not understood by terra code.
    add1(&a)
       ^
a.t:7: to call a lua function from terra first use terralib.cast to cast it to a terra function type.
    add1(&a)
       ^
a.t:7: expected a function or macro but found lua value of type function
    add1(&a)
        ^

stack traceback:
	[C]: in function 'error'
	/terra/src/terralib.lua:396: in function 'finishandabortiferrors'
	/terra/src/terralib.lua:3431: in function 'typecheck'
	/terra/src/terralib.lua:1172: in function 'defineobjects'
	a.t:4: in main chunk
```

I tried a few things here but nothing made it work. Is the example wrong?

**Comment by elliottslaughter** [comment]
This behavior was actually changed way back in [release 2016-03-25](https://github.com/terralang/terra/releases/tag/release-2016-03-25) (see under the heading "Calling Lua Functions from Terra"). I guess the example never got updated to match.

Please try this version:

```terra
struct Complex { real : double, imag : double }
local add1 = terralib.cast({&Complex}->{},function(a)
    a.real = a.real + 1
end)
terra doit()
    var a = Complex {1,2}
    add1(&a)
    return a
end
a = doit()
print(a.real,a.imag) -- 2    1
print(type(a)) -- cdata
```

I'll keep this issue open until the docs have been fixed.

**Comment by renatoathaydes** [comment]
Yeah that worked. I couldn't figure out the syntax for casting that the error message suggested.
Hope you can update the examples, closing this issue.

**Comment by elliottslaughter** [comment]
This should be fixed in 9b06b822e1f0d4a3f849751d2476a12b8cd56b1f.
